### PR TITLE
Adds Recitation 8

### PR DESCRIPTION
**Pull request by vasumv**
Adds Recitation 8 for LLMs.

- [x] Add Handout (new recitation)
- [x] Update Course Schedule

### CODE REVIEW

File: docs/recitations/reci8-llms.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+1.  Enable the Vertex AI API using [this link](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com). Make sure your NodeBB project from P3 is selected in the project selector dropdown.
+    
+2.  Install JupyterLab for Python following the instructions in [this link](https://jupyter.org/install).
+    
+3.  Install the gcloud CLI following the instructions in [this link](https://cloud.google.com/sdk/docs/install).
+    
+4.  In your terminal, run `gcloud auth login` and open the authentication URL that is displayed. Choose the account linked to the NodeBB project, and copy the verification code and paste it into the terminal.
+    
+5.  In your terminal, run `gcloud config set project PROJECT_ID`, where the `PROJECT_ID` is the ID of your NodeBB project.
```
Comment (anudaweerasinghe) [inline docs/recitations/reci8-llms.md:17]: Where can they find this `PROJECT_ID`?

File: docs/recitations/reci8-llms.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+## Overview
+
+During this recitation, students will able to run and evaluate a large language model on an example dataset. This is an individual activity so everyone should be running the notebook; however, feel free to discuss the activity and your approach with other students!
+
+## Activity 1: Examine the Stanford Question Answering (SQuAD) Dataset (10 mins)
```
Comment (anudaweerasinghe) [inline docs/recitations/reci8-llms.md:23]: Change the titles on the notebook to match these headings.

File: docs/recitations/reci8-llms.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+3.  Install the gcloud CLI following the instructions in [this link](https://cloud.google.com/sdk/docs/install).
+    
+4.  In your terminal, run `gcloud auth login` and open the authentication URL that is displayed. Choose the account linked to the NodeBB project, and copy the verification code and paste it into the terminal.
+    
+5.  In your terminal, run `gcloud config set project PROJECT_ID`, where the `PROJECT_ID` is the ID of your NodeBB project.
+
+## Overview
+
+During this recitation, students will able to run and evaluate a large language model on an example dataset. This is an individual activity so everyone should be running the notebook; however, feel free to discuss the activity and your approach with other students!
+
```
Comment (anudaweerasinghe) [inline docs/recitations/reci8-llms.md:22]: Include an "Activity 0" where people download the notebook, and run the "Install Python Dependencies" section

File: docs/recitations/reci8-llms.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+## Activity 1: Examine the Stanford Question Answering (SQuAD) Dataset (10 mins)
+
+The dataset contains a set of questions and answers to test a system for reading comprehension. Download the notebook from [this repository](https://github.com/CMU-313/fa23-recitation-8) and go through the first part of the Jupyter notebook to install any Python dependencies and load in the dataset. 
+
+For one example, print out the context, question, and expected answer.
```
Comment (anudaweerasinghe) [inline docs/recitations/reci8-llms.md:27]: This seems to be already given, do you plan on removing it from the notebook?

File: docs/recitations/reci8-llms.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+## Setup Instructions (10 min)
+For this recitation and Project 4, you'll be using Google's Vertex AI Platform in a Jupyter notebook to query a large language model. First, we need to set up our GCP project to enable the API and install the prerequisites to run the notebook. 
+
+1.  Enable the Vertex AI API using [this link](https://console.cloud.google.com/flows/enableapi?apiid=aiplatform.googleapis.com). Make sure your NodeBB project from P3 is selected in the project selector dropdown.
+    
+2.  Install JupyterLab for Python following the instructions in [this link](https://jupyter.org/install).
+    
+3.  Install the gcloud CLI following the instructions in [this link](https://cloud.google.com/sdk/docs/install).
+    
+4.  In your terminal, run `gcloud auth login` and open the authentication URL that is displayed. Choose the account linked to the NodeBB project, and copy the verification code and paste it into the terminal.
```
Comment (anudaweerasinghe) [inline docs/recitations/reci8-llms.md:15]: I had to run `gcloud auth application-default login` to fix auth problems in cell containing `query_llm`.

Note: I used homebrew to install the gcloud cli

**Review by anudaweerasinghe — CHANGES_REQUESTED** [review]
Thanks Vasu, colab setup is much cleaner. Had no problems with setup/running the notebook.

Left some comments about dataset, and eval implementation.

File: docs/recitations/reci8-llms.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+## Activity 2: Use the Vertex AI API to query the chat-bison LLM (5 min)
+
+Now that we have loaded the dataset, let's try querying an LLM to answer one of the questions. Refer to the [Vertex Python SDK docs](https://cloud.google.com/vertex-ai/docs/generative-ai/chat/test-chat-prompts#chat-query-python_vertex_ai_sdk) to load and query the model. For this exercise, choose the "**chat-bison@001**" model. 
+
+Try running the model on one of the examples from the SQuAD dataset. Does the output match the expected answer?
+
+## Activity 3: Check Correctness of Output (10 min)
+
+When using an LLM, it is important to design an evaluation strategy for the task. This involves choosing a method of evaluating the correctness of the LLM output. Since the output in our task is textual, there are many difference ways of checking similarity to an expected answer. For this recitation,
+we will use [_cosine similarity_](https://en.wikipedia.org/wiki/Cosine_similarity).
```
[the remark this answers is not in the document]
Comment (anudaweerasinghe) [reply docs/recitations/reci8-llms.md:38]: Using [this](https://www.sbert.net/docs/quickstart.html#comparing-sentence-similarities) might actually be better. It's a much cleaner abstraction of computing cosine similarities on generated embeddings. We can use 'sentence-transformers/msmarco-distilbert-base-tas-b' as a the model.

File: docs/recitations/reci8-llms.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+During this recitation, students will able to run and evaluate a large language model on an example dataset. This is an individual activity so everyone should be running the notebook; however, feel free to discuss the activity and your approach with other students!
+
+## Activity 0: Install Dependencies and Authenticate GCP (5 min)
+
+Run the first cell of the notebook to install any Python dependencies. Follow the rest of the instructions in the notebook to authenticate GCP.
+
+## Activity 1: Examine the Stanford Question Answering (SQuAD) Dataset (10 mins)
+
+The dataset contains a set of questions and answers to test a system for reading comprehension. Load in the dataset and understand the format.
```
[the remark this answers is not in the document]
Comment (vasumv) [reply docs/recitations/reci8-llms.md:25]: good call

File: docs/recitations/reci8-llms.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (vasumv) [reply docs/recitations/reci8-llms.md:38]: Changed to use "contains" for this recitation, and added links to bert-score, sbert, GCP text embeddings.

**Review by anudaweerasinghe — APPROVED** [review]

**Pull request merged**